### ISSUE TITLE
dsolve issues with coupled underdamped first order ODEs

### ISSUE THREAD

**Issue by DukeDoc99** [sympy/sympy#22715]
Bottom Line Up Front:  It looks like dsolve does not properly solve two coupled linear first-order differential equations if the system is underdamped (eigenvalues are complex).  It works great for other damping (un, over, critical) and also works if the system is converted into a single second-order underdamped differential equation.

Also - apologies in advance that this is not formatted well.  I am off in search of how to better communicate on GitHub threads!

I was looking at solving a system of equations that has the general form:

[𝑑𝑑𝑡𝑥(𝑡)=𝑎𝑥(𝑡)+𝑏𝑦(𝑡)+𝑐, 𝑑𝑑𝑡𝑦(𝑡)=𝑑𝑥(𝑡)+𝑒𝑦(𝑡)+𝑓]

and it would appear that in cases where the eigenvalues are complex (underdamped), the dsolve system gives either an overcomplicated answer (if f=0) or...a complete disaster (if f is not 0).  It works great for real, real repeated, and purely imaginary eigenvalues.  I have a function called eqnsubs to help make substitutions in equations.  Here's the full notebook:

[Damping Issue Demo](https://github.com/DukeDoc99/AS7e/blob/main/DampingDEdemo.ipynb)

And here's the start with a couple examples

```
import sympy as sym
sym.init_printing()

def eqnsubs(eqns, vartosub, valtosub):
    subeqn = eqns.copy()
    sublist = list(zip(vartosub, valtosub))
    for idx, eqn in enumerate(subeqn):
        subeqn[idx] = subeqn[idx].subs(sublist)
    return subeqn 

sym.var('a b c d e f t')
sym.var('x y', cls=sym.Function)

eqn1 = sym.Eq(sym.diff(x(t),t),a*x(t)+b*y(t)+c)
eqn2 = sym.Eq(sym.diff(y(t),t),d*x(t)+e*y(t)+f)
eqns = [eqn1, eqn2]
eqns
```

and here's the part showing where underdamped and overdamped diverge:

```
# Overdamped - works great
eqnso = eqnsubs(eqns, (a, b, c, d, e, f), (0, -1, 10, 4, -5, 3))
solno = sym.dsolve(eqnso)
solno
```
yields
[𝑥(𝑡)=𝐶1 exp(−𝑡) + 𝐶2 exp(−4𝑡) / 4 + 47 / 4, 𝑦(𝑡)=𝐶1 exp(−𝑡) + 𝐶2 exp(−4𝑡) +10]

whereas:

```
eqnsub = eqnsubs(eqns, (a, b, c, d, e, f), (0, -1, 10, 5, -2, 3))
solnub = sym.dsolve(eqnsub)
solnub
```

yields...erm...

𝑥(𝑡)=(𝐶15−2𝐶25)𝑒−𝑡cos(2𝑡)−(2𝐶15+𝐶25)𝑒−𝑡sin(2𝑡)+2sin2(2𝑡)+47sin(2𝑡)cos(2𝑡)10+7cos2(2𝑡)5+47𝑒−𝑡sin(2𝑡)∫𝑒𝑡sin(2𝑡)𝑑𝑡5−6𝑒−𝑡sin(2𝑡)∫𝑒𝑡cos(2𝑡)𝑑𝑡5+6𝑒−𝑡sin(2𝑡)∫𝑒𝑡sin2(2𝑡)cos(2𝑡)𝑑𝑡5−47𝑒−𝑡cos(2𝑡)∫𝑒𝑡sin(2𝑡)𝑑𝑡10+3𝑒−𝑡cos(2𝑡)∫𝑒𝑡cos(2𝑡)𝑑𝑡5−3𝑒−𝑡cos(2𝑡)∫𝑒𝑡sin2(2𝑡)cos(2𝑡)𝑑𝑡5, 𝑦(𝑡)=𝐶1𝑒−𝑡cos(2𝑡)−𝐶2𝑒−𝑡sin(2𝑡)+10sin2(2𝑡)+7sin(2𝑡)cos(2𝑡)2−47𝑒−𝑡cos(2𝑡)∫𝑒𝑡sin(2𝑡)𝑑𝑡2+3𝑒−𝑡cos(2𝑡)∫𝑒𝑡cos(2𝑡)𝑑𝑡−3𝑒−𝑡cos(2𝑡)∫𝑒𝑡sin2(2𝑡)cos(2𝑡)𝑑𝑡

(trust me - it isn't pretty).

**Comment by oscarbenjamin** [comment]
The formula here looks incorrect:
https://github.com/sympy/sympy/blob/60b926dffabf0801bb6c7eb91168a3d175b74dd3/sympy/solvers/ode/systems.py#L983
The idea is that we have a first order matrix system of the form
```
dX/dt = A*X + b     (1)
```
where x and b are column vectors and A is a matrix. In context the equations are:
```
In [7]: eqnsub
Out[7]: 
⎡d                     d                             ⎤
⎢──(x(t)) = 10 - y(t), ──(y(t)) = 5⋅x(t) - 2⋅y(t) + 3⎥
⎣dt                    dt                            ⎦
```
So `X = Matrix([x, y])` and `A = Matrix([[0, -1], [5, -2]])` and `b = Matrix([10, 3])`.

For an equation of the form (1) we can construct the solution analogous to the integrating factor method for first order single ODEs. We first subtract `A*X` to the left and then multiply by `exp(-A*t)` (matrix exponential):
```
exp(-A t) dX/dt - exp(-A t) A X = exp(-A t) b
```
Then we recognise that the lhs is like the product rule:
```
d/dt { exp(-A t) X } = exp(-A t) b
```
Now integrate:
```
exp(-A t) X = Integral(exp(-A t) b, t) + C
```
Now the matrix exponential of A is computed from the Jordan normal form of A so `A = P_A J_A P_A**-1` and `exp(A t) = P_A exp(J_A t) P_A**-1`. Hence the inverse `exp(-A t)` is `P_A exp(J_A t)**-1 P_A**-1`. Here `P` and `J` are such that `exp(A t) = P J P**-1` so `exp(-A t) = P J**-1 P` and we have
```
P J**-1 P X = Integral(P J**-1 P b, t) + C
```
We multiply through by `P**-1` and absorb that into C to get
```
J**-1 P X = Integral(J**-1 P b, t) + C'
```
Now invert the matrices on the lhs and we have
```
X = P**-1 J (Integral(J**-1 P b, t) + C')
```
which is different from the formula shown above in the code. Not sure if I've missed something here...

**Comment by oscarbenjamin** [comment]
CC @mijo2

**Comment by DukeDoc99** [comment]
@oscarbenjamin - 

I am not sure about 983 (it may be an alternate form - I agree with your formulation) and the calculations end up as transposes of each other.  I went through and tried out some of those calculations with the under and overdamped matrix and what I am finding is that the integral remains undone for underdamped (even in the presence of doit):

```
import sympy as sym
sym.var('t, con', real=True)
sym.init_printing()

b = sym.Matrix([10, 3])

Au = sym.Matrix([[0, -1], [5, -2]])
Pu, Ju = sym.solvers.ode.systems.matrix_exp_jordan_form(Au, t)
Pu = sym.simplify(Pu)
print(Pu,"\n")
print(Ju,"\n")
print(Ju.inv() * Pu.inv(),"\n")
Intu = Pu * Ju * (sym.Integral(Ju.inv() * Pu.inv() * b, t)+con).doit()
print(Intu,"\n")
```
leaves the unevaluated integrals
```
Matrix([[(-2*exp(-t)*sin(2*t)/5 + exp(-t)*cos(2*t)/5)*(con + Matrix([
[-(Integral(47*exp(t)*sin(2*t), t) + Integral(-6*exp(t)/cos(2*t), t) + Integral(6*exp(t)*sin(2*t)**2/cos(2*t), t))/2],
[                                                                          -10*exp(t)*sin(2*t) - 7*exp(t)*cos(2*t)/2]])), (-exp(-t)*sin(2*t)/5 - 2*exp(-t)*cos(2*t)/5)*(con + Matrix([
[-(Integral(47*exp(t)*sin(2*t), t) + Integral(-6*exp(t)/cos(2*t), t) + Integral(6*exp(t)*sin(2*t)**2/cos(2*t), t))/2],
[                                                                          -10*exp(t)*sin(2*t) - 7*exp(t)*cos(2*t)/2]]))], [exp(-t)*cos(2*t)*(con + Matrix([
[-(Integral(47*exp(t)*sin(2*t), t) + Integral(-6*exp(t)/cos(2*t), t) + Integral(6*exp(t)*sin(2*t)**2/cos(2*t), t))/2],
[                                                                          -10*exp(t)*sin(2*t) - 7*exp(t)*cos(2*t)/2]])), -exp(-t)*sin(2*t)*(con + Matrix([
[-(Integral(47*exp(t)*sin(2*t), t) + Integral(-6*exp(t)/cos(2*t), t) + Integral(6*exp(t)*sin(2*t)**2/cos(2*t), t))/2],
[                                                                          -10*exp(t)*sin(2*t) - 7*exp(t)*cos(2*t)/2]]))]]) 
```

at the end while
```
Ao = sym.Matrix([[0, -1], [4, -5]])
Po, Jo = sym.solvers.ode.systems.matrix_exp_jordan_form(Ao, t)
Po = sym.simplify(Po)
print(Po,"\n")
print(Jo,"\n")
print(Jo.inv() * Po.inv(),"\n")
Into = Po * Jo * (sym.Integral(Jo.inv() * Po.inv() * b, t)+con).doit()
print(Into)
```

leaves no integrals

```
Matrix([[exp(-4*t)*(con + Matrix([
[-7*exp(4*t)/3],
[  37*exp(t)/3]]))/4, exp(-t)*(con + Matrix([
[-7*exp(4*t)/3],
[  37*exp(t)/3]]))], [exp(-4*t)*(con + Matrix([
[-7*exp(4*t)/3],
[  37*exp(t)/3]])), exp(-t)*(con + Matrix([
[-7*exp(4*t)/3],
[  37*exp(t)/3]]))]])
```

I am trying to see if there is a way to may it determine the integrals - I am not having luck getting it to make any sense out of:
```
Q = sym.Integral(-6*sym.exp(t)/sym.cos(2*t), t)
```
for example...

**Comment by DukeDoc99** [comment]
So - looks like the issue might be in taking the inverse of J and having trig in the denominator - if I use 
```
Intu = Pu * Ju * (sym.Integral(sym.simplify(Ju.inv() * Pu.inv() * b), t)+con).doit()
print(Intu,"\n")
```
the Integrals go away...pursuing further!

**Comment by DukeDoc99** [comment]
Replacing line 983 in system.py with

```
            sol_vector = P * J * ((simplify(J.inv()) * P.inv() * b).applyfunc(lambda x: Integral(x, t)) + Cvect)
```

allows underdamped systems to get evaluated answers!  Not sure what that might break (and also, the answers are still coming out with interesting trig like cos^2 + sin^2 not being simplified to 1) but it makes my example files work  :)

**Comment by oscarbenjamin** [comment]
We don't usually want to call simplify in library code. I see though that simplifying the inverse does lead to the expected answer. In this case maybe computing the simplified inverse as as simple as negating `t` though:
```python
In [4]: J = Matrix([
   ...: [exp(-t)*cos(2*t), -exp(-t)*sin(2*t)],
   ...: [exp(-t)*sin(2*t),  exp(-t)*cos(2*t)]])

In [5]: J
Out[5]: 
⎡ -t             -t         ⎤
⎢ℯ  ⋅cos(2⋅t)  -ℯ  ⋅sin(2⋅t)⎥
⎢                           ⎥
⎢ -t            -t          ⎥
⎣ℯ  ⋅sin(2⋅t)  ℯ  ⋅cos(2⋅t) ⎦

In [6]: J.inv()
Out[6]: 
⎡   t    2            t                ⎤
⎢  ℯ ⋅sin (2⋅t)      ℯ       t         ⎥
⎢- ──────────── + ────────  ℯ ⋅sin(2⋅t)⎥
⎢    cos(2⋅t)     cos(2⋅t)             ⎥
⎢                                      ⎥
⎢        t                   t         ⎥
⎣      -ℯ ⋅sin(2⋅t)         ℯ ⋅cos(2⋅t)⎦

In [7]: simplify(J.inv())
Out[7]: 
⎡ t             t         ⎤
⎢ℯ ⋅cos(2⋅t)   ℯ ⋅sin(2⋅t)⎥
⎢                         ⎥
⎢  t            t         ⎥
⎣-ℯ ⋅sin(2⋅t)  ℯ ⋅cos(2⋅t)⎦

In [8]: J.subs(t, -t)
Out[8]: 
⎡ t             t         ⎤
⎢ℯ ⋅cos(2⋅t)   ℯ ⋅sin(2⋅t)⎥
⎢                         ⎥
⎢  t            t         ⎥
⎣-ℯ ⋅sin(2⋅t)  ℯ ⋅cos(2⋅t)⎦
```
So maybe we just do `Jinv = J.subs(t, -t)` and use `Jinv` instead of `J.inv()`

**Comment by oscarbenjamin** [comment]
This seems to do the trick:
```diff
diff --git a/sympy/solvers/ode/systems.py b/sympy/solvers/ode/systems.py
index a31911c..96a88fd 100644
--- a/sympy/solvers/ode/systems.py
+++ b/sympy/solvers/ode/systems.py
@@ -980,7 +980,8 @@ def linodesolve(A, t, b=None, B=None, type="auto", doit=False,
         if type in ("type1", "type5"):
             sol_vector = P * (J * Cvect)
         else:
-            sol_vector = P * J * ((J.inv() * P.inv() * b).applyfunc(lambda x: Integral(x, t)) + Cvect)
+            Jinv = J.subs(t, -t)
+            sol_vector = P * J * ((Jinv * P.inv() * b).applyfunc(lambda x: Integral(x, t)) + Cvect)
 
     else:
         if B is None:
```
With that:
```python
In [1]: import sympy as sym
   ...: sym.init_printing()
   ...: 
   ...: def eqnsubs(eqns, vartosub, valtosub):
   ...:     subeqn = eqns.copy()
   ...:     sublist = list(zip(vartosub, valtosub))
   ...:     for idx, eqn in enumerate(subeqn):
   ...:         subeqn[idx] = subeqn[idx].subs(sublist)
   ...:     return subeqn
   ...: 
   ...: sym.var('a b c d e f t')
   ...: sym.var('x y', cls=sym.Function)
   ...: 
   ...: eqn1 = sym.Eq(sym.diff(x(t),t),a*x(t)+b*y(t)+c)
   ...: eqn2 = sym.Eq(sym.diff(y(t),t),d*x(t)+e*y(t)+f)
   ...: eqns = [eqn1, eqn2]
   ...: eqns
Out[1]: 
⎡d                               d                             ⎤
⎢──(x(t)) = a⋅x(t) + b⋅y(t) + c, ──(y(t)) = d⋅x(t) + e⋅y(t) + f⎥
⎣dt                              dt                            ⎦

In [2]: eqnsub = eqnsubs(eqns, (a, b, c, d, e, f), (0, -1, 10, 5, -2, 3))
   ...: solnub = sym.dsolve(eqnsub)
   ...: solnub
Out[2]: 
⎡                                                                   2              2               
⎢       ⎛C₁   2⋅C₂⎞  -t            ⎛2⋅C₁   C₂⎞  -t            17⋅sin (2⋅t)   17⋅cos (2⋅t)          
⎢x(t) = ⎜── - ────⎟⋅ℯ  ⋅cos(2⋅t) - ⎜──── + ──⎟⋅ℯ  ⋅sin(2⋅t) + ──────────── + ────────────, y(t) = C
⎣       ⎝5     5  ⎠                ⎝ 5     5 ⎠                     5              5                

                                                              ⎤
   -t                -t                  2              2     ⎥
₁⋅ℯ  ⋅cos(2⋅t) - C₂⋅ℯ  ⋅sin(2⋅t) + 10⋅sin (2⋅t) + 10⋅cos (2⋅t)⎥
                                                              ⎦

In [3]: checkodesol(eqnsub, solnub)
Out[3]: (True, [0, 0])
```

**Comment by oscarbenjamin** [comment]
But maybe that's only valid for `type2`. In the case of `type6` we might need something else as well...

**Comment by DukeDoc99** [comment]
>  We don't usually want to call simplify in library code.

Gotcha - although I will mention there is a "P = simplify(P)" in line 978  :)   

Also - thank you for all your work on this specifically and the project generally!

**Comment by oscarbenjamin** [comment]
> Gotcha - although I will mention there is a "P = simplify(P)" in line 978 :)

Yes, I saw that when I actually looked at the code again. I don't think it should be there. I tried removing it and running the tests like
```console
$ pytest sympy/solvers/ode/tests/test_systems.py -m 'not slow'
```
There were a few failures e.g.:
```python
In [3]: C1, C2 = symbols('C1, C2')

In [4]: x, y = symbols('x, y', cls=Function)

In [5]:         eqs9 = [Eq(Derivative(x(t), t), x(t) + y(t)),
   ...:                 Eq(Derivative(y(t), t), x(t) - y(t))]
   ...:         sol9 = [Eq(x(t), C1*(1 - sqrt(2))*exp(-sqrt(2)*t) + C2*(1 + sqrt(2))*exp(sqrt(2)*t)),
   ...: 
   ...:                 Eq(y(t), C1*exp(-sqrt(2)*t) + C2*exp(sqrt(2)*t))]

In [6]: eqs9
Out[6]: 
⎡d                       d                     ⎤
⎢──(x(t)) = x(t) + y(t), ──(y(t)) = x(t) - y(t)⎥
⎣dt                      dt                    ⎦

In [7]: sol9
Out[7]: 
⎡                    -√2⋅t                √2⋅t             -√2⋅t       √2⋅t⎤
⎣x(t) = C₁⋅(1 - √2)⋅ℯ      + C₂⋅(1 + √2)⋅ℯ    , y(t) = C₁⋅ℯ      + C₂⋅ℯ    ⎦

In [8]: dsolve(eqs9)
Out[8]: 
⎡                    √2⋅t                -√2⋅t             √2⋅t       -√2⋅t⎤
⎣x(t) = C₁⋅(1 + √2)⋅ℯ     + C₂⋅(1 - √2)⋅ℯ     , y(t) = C₁⋅ℯ     + C₂⋅ℯ     ⎦

In [9]: checkodesol(eqs9, sol9)
Out[9]: (True, [0, 0])

In [10]: checkodesol(eqs9, dsolve(eqs9))
Out[10]: (True, [0, 0])
```
In that case it seems that the test could just be changed since both answers are equally correct and similarly simple. All that's happened is that the constants C1 and C2 have been swapped. I'm not sure if all the other cases are similar but maybe it is possible to just remove the call to `simplify` there.

**Comment by DukeDoc99** [comment]
In case you are truly bored - I had mentioned the issue to some of my colleagues and wanted to document the process so here's that:
https://github.com/DukeDoc99/Development/blob/main/jordan_problems.ipynb

**Comment by oscarbenjamin** [comment]
Thanks for that. Looking back at my original post above I see where the mistake was that I wrote `P` instead of `P**-1` somewhere that carried through.

Since I've attached a diff that fixes the issue I'm going to mark this as easy to fix. If anyone wants to put that diff plus a test for the example in the OP into a PR then go ahead.

Another improvement that would be nice would be removing the call to `simplify` but it needs a bit of investigation to see whether there are some cases where the output is significantly less good.
